torch/distributed/distributed_c10d.py

```diff
@@ -1260,17 +1282,37 @@ def _new_process_group_helper(
     # The list of group ranks is empty if we're creating the default group.
     is_default_group = len(global_ranks_in_group) == 0
 
+    # nccl and potentially other backends allow creation of
+    # communicators based on pre-existing ones, which can save
+    # initialization time.  Due to lazy initialization of
+    # communicators in some backends, we have to be careful and only
+    # split when we *know* the backends already are connected _on all
+    # ranks_.  We can only know this if the group we are making is the
+    # entire world or if we have bound a device id to the world (which
+    # causes early connection initialization).
+    if (is_initialized() and
+            (len(global_ranks_in_group) == _world.default_pg.size() or _world.default_pg.bound_device_id)):
+        should_split = True
+        split_from = _get_split_source(_world.default_pg)
+    else:
+        should_split = False
+        split_from = None
+
     # If this is a subgroup (which means group_ranks is specified),
     # we check if the current process is a member of the new group.
     if not is_default_group:
         global_rank = _get_default_group().rank()
         if global_rank not in global_ranks_in_group:
+            if split_from:
+                split_from.perform_nocolor_split(_world.default_pg.bound_device_id)
```

Comment:
If the current rank is not in this sub-pg, why we still want to have this logic here?

Reply:
`ncclCommSplit` must be called on all communicators we're splitting from, even ones not in the new subgroup.  So we must call it here when we aren't one of the global ranks in the group.  This one small requirement is what has made this change complex: lazy creation of communicators makes it impossible to call the nocolor split unless we do it immediately (otherwise ranks not in the new group may not call it in sync with the ones that do regular splits).

Sounds like I should comment this better.  Where would you recommend?  Here?

Reply:
For learning purpose, so all ranks have to call it and the ranks which are not in the group should just call the `perform_nocolor_split`? I see, that's why you have rank guess logic because if we do lazy init, when we do the split, the communicator might have not been initialized?

Reply:
For the comment, I think this is a good place because this is a common helper function called by multiple public APIs.

Reply:
Yes exactly.  Ranks that don't join the new group still must call it.  I'll add a comment here.  If we were doing lazy init we wouldn't know when or if the non-member ranks might (ever) call a collective function.  And we need them to call a nocolor split before any collective operation.  I'll add a comment.